src/plopp/graphics/color_mapper.py

```diff
@@ -40,52 +45,111 @@ def __init__(self,
         self.mask_cmap = _get_cmap(mask_cmap, nan_color=nan_color)
         self.user_vmin = vmin
         self.user_vmax = vmax
-        self.vmin = np.inf
-        self.vmax = np.NINF
-        self.norm_flag = norm
-        self.norm_func = None
+        self._vmin = {'linear': np.inf, 'log': np.inf}
+        self._vmax = {'linear': np.NINF, 'log': np.NINF}
+        self._norm = norm
+        self.normalizer = {'linear': Normalize(), 'log': LogNorm()}
+        self._notify_on_change = [notify_on_change]
+        self.colorbar = None
+        self.unit = None
+        self.name = None
 
-    def rescale(self, data):
+    @property
+    def widget(self):
+        import ipywidgets as ipw
+        self.colorbar = {
+            'image':
+            ipw.Image(),
+            'button':
+            ToggleTool(self.toggle_norm,
+                       value=self._norm == 'log',
+                       description='log',
+                       tooltip='Toggle data norm').widget
+        }
+        self._update_colorbar()
+        return ipw.VBox(list(self.colorbar.values()))
+
+    def _update_colorbar(self):
+        dpi = 96
+        height_px = 600
+        height_inches = 0.89 * height_px / dpi
+        cbar_fig = plt.figure(figsize=(height_inches * 0.2, height_inches), dpi=dpi)
+        cbar_ax = cbar_fig.add_axes([0.05, 0.02, 0.25, 1.0])
+        _ = ColorbarBase(cbar_ax, cmap=self.cmap, norm=self.norm)
+        cbar_ax.set_ylabel(f'{self.name} [{self.unit}]')
+        self.colorbar['image'].value = fig_to_bytes(cbar_fig)
+        plt.close(cbar_fig)
+
+    def autoscale(self, data, scale):
         """
         Re-compute the min and max range of values, given new values.
         """
-        vmin, vmax = fix_empty_range(find_limits(data, scale=self.norm_flag))
+        vmin, vmax = fix_empty_range(find_limits(data, scale=scale))
         if self.user_vmin is not None:
             assert self.user_vmin.unit == data.unit
-            self.vmin = self.user_vmin.value
-        elif vmin.value < self.vmin:
-            self.vmin = vmin.value
+            self._vmin[scale] = self.user_vmin.value
+        elif vmin.value < self._vmin[scale]:
+            self._vmin[scale] = vmin.value
         if self.user_vmax is not None:
             assert self.user_vmax.unit == data.unit
-            self.vmax = self.user_vmax.value
-        elif vmax.value > self.vmax:
-            self.vmax = vmax.value
+            self._vmax[scale] = self.user_vmax.value
+        elif vmax.value > self._vmax[scale]:
+            self._vmax[scale] = vmax.value
 
-        self.norm_func.vmin = self.vmin
-        self.norm_func.vmax = self.vmax
+        self.normalizer[scale].vmin = self._vmin[scale]
+        self.normalizer[scale].vmax = self._vmax[scale]
```

Comment:
This only scales for the given `scale`. When the plot is, e.g. in linear mode and it is auto scaled, and the user then switches to log mode, does it use the scale that was previously used with log? Is this what we want?

Reply:
This was my way of not needing to request the data again when toggling the log. You compute the limits for both linear and log when you first make the figure, and store those. But it is true that it will use old limits if you then move a slider and change the colors, and then switch to log, your limits will be out of date (I think?)... I should probably change.